### ISSUE TITLE
Multi GPU process stuck in the dataloader loop, single GPU works fine.

### ISSUE THREAD

**Issue by SHEN2BAIYI** [huggingface/accelerate#2768]
### System Info

```Shell
- `Accelerate` version: 0.30.0
- Platform: Linux-5.11.0-41-generic-x86_64-with-glibc2.31
- `accelerate` bash location: /home/zijieshen/anaconda3/envs/pySEG/bin/accelerate
- Python version: 3.9.19
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.2.0+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 62.50 GB
- GPU type: NVIDIA GeForce RTX 3090
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: no
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- rdzv_backend: static
	- same_network: False
	- main_training_function: main
	- enable_cpu_affinity: False
	- downcast_bf16: False
	- tpu_use_cluster: False
	- tpu_use_sudo: False
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

import torch
import torchmetrics
from accelerate import notebook_launcher
from torchkeras.kerasmodel import KerasModel

# 设置 NCCL_P2P_DISABLE 环境变量
import os
os.environ["NCCL_P2P_DISABLE"] = "1"
os.environ["NCCL_IB_DISABLE"] = "1"
### 1，准备数据

def create_dataloaders(batch_size=2):
    transform = transforms.Compose([transforms.ToTensor()])

    ds_train = torchvision.datasets.MNIST(root="./minist/",train=True,download=True,transform=transform)
    ds_val = torchvision.datasets.MNIST(root="./minist/",train=False,download=True,transform=transform)

    dl_train =  torch.utils.data.DataLoader(ds_train, batch_size=batch_size, shuffle=True,
                                            num_workers=2,drop_last=True)
    dl_val =  torch.utils.data.DataLoader(ds_val, batch_size=batch_size, shuffle=False,
                                          num_workers=2,drop_last=True)
    return dl_train,dl_val

dl_train,dl_val = create_dataloaders(batch_size=64)

### 2，定义模型

def create_net():
    net = nn.Sequential()
    net.add_module("conv1",nn.Conv2d(in_channels=1,out_channels=512,kernel_size = 3))
    net.add_module("pool1",nn.MaxPool2d(kernel_size = 2,stride = 2))
    net.add_module("conv2",nn.Conv2d(in_channels=512,out_channels=256,kernel_size = 5))
    net.add_module("pool2",nn.MaxPool2d(kernel_size = 2,stride = 2))
    net.add_module("dropout",nn.Dropout2d(p = 0.1))
    net.add_module("adaptive_pool",nn.AdaptiveMaxPool2d((1,1)))
    net.add_module("flatten",nn.Flatten())
    net.add_module("linear1",nn.Linear(256,128))
    net.add_module("relu",nn.ReLU())
    net.add_module("linear2",nn.Linear(128,10))
    return net

net = create_net()

### 3，训练模型

loss_fn = nn.CrossEntropyLoss()
metrics_dict = {'acc':torchmetrics.Accuracy(task='multiclass',num_classes=10)}

optimizer = torch.optim.AdamW(params=net.parameters(), lr=1e-4)
lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingWarmRestarts(
    optimizer=optimizer,T_0=5)

model = KerasModel(net,loss_fn,metrics_dict,optimizer,lr_scheduler)

ckpt_path = 'checkpoint.pt'
args = dict(train_data = dl_train,
        val_data = dl_val,
        epochs=10,
        ckpt_path= ckpt_path,
        patience=2,
        monitor='val_acc',
        mode='max',
        callbacks=None,
        plot=False,
        wandb=False,
        quiet=None,
        mixed_precision='no'
)

model.fit_ddp(
    num_processes=2,
    **args
)

### Expected behavior

A normal training process.

**Comment by SHEN2BAIYI** [comment]
When I use debug mode of Pycharm, I find that the procedure will stuck in the dataloader loop. 


![微信图片_20240513190551](https://github.com/huggingface/accelerate/assets/96118084/7c9d1442-e908-4f26-942f-3a780b9dc6c9)

**Comment by muellerzr** [comment]
You may need to set the env variable inside the training function. Can you share your full notebook code?

**Comment by SHEN2BAIYI** [comment]
> You may need to set the env variable inside the training function. Can you share your full notebook code?

Thank you for your prompt reply, it doesn't work as well. And the code above is all I have. I'm trying to standardize my code with torchkeras.

**Comment by muellerzr** [comment]
Not sure if we support `TorchKeras` models. Does `isinstance(model, torch.nn.Module)` return `True`?

**Comment by SHEN2BAIYI** [comment]
> isinstance(model, torch.nn.Module)

yes, it returns Ture. And I think we can think of torchkeras as a normal training program.

**Comment by muellerzr** [comment]
Why are we combining accelerate then? I'm not sure those two can exist at the same time. This feels like a `torchkeras` issue more than accelerate.

**Comment by SHEN2BAIYI** [comment]
> Why are we combining accelerate then? I'm not sure those two can exist at the same time. This feels like a `torchkeras` issue more than accelerate.

Thanks, I find the solution...This is a bug of Pytorch. In 'torch.utils.data.sampler', we must use "torch.randperm" method, which does not work in a multiprocess environment when passed a value greater than 2**15. So I reduce the size of my dataset, and it works!!!
